include/jemalloc/internal/cache_bin.h

```diff
@@ -427,6 +428,30 @@ cache_bin_full(cache_bin_t *bin) {
 	return ((uint16_t)(uintptr_t)bin->stack_head == bin->low_bits_full);
 }
 
+/*
+ * Scans the allocated area of the cache_bin for the given pointer up to limit.
+ * Fires safety_check_fail if the ptr is found and returns false.
+ */
+JEMALLOC_ALWAYS_INLINE bool
+cache_bin_validate_double_free(cache_bin_t *bin, void *ptr) {
+        void **cur = bin->stack_head;
+        unsigned limit = opt_debug_dblfree_cbin_scan_limit;
+
+        while (limit != 0 && ((uint16_t)(uintptr_t)cur != bin->low_bits_full)
+            && *cur != ptr) {
+                cur++;
+                limit--;
+        }
+
+        // ptr is already present in the stack
+        if (limit != 0 && ((uint16_t)(uintptr_t)cur != bin->low_bits_full)) {
+                safety_check_fail(
+                    "Invalid deallocation detected: double free\n");
+                return true;
+        }
```

Comment:
With the `max_scan == 0` check at the beginning, we can simplify above a bit:
```
 while (limit != 0 && (uint16_t)(uintptr_t)cur != bin->low_bits_full) {
    if (*cur == ptr) {
        ...
    }
    limit--;
    cur++;
}
```

Reply:
very elegant!

Reply:
addressed in [a903674](https://github.com/jemalloc/jemalloc/pull/2310/commits/a903674db54ec8e8fc00b588e4e4279b0e0cd965)